### PR TITLE
[nrfconnect] Include GenericSwitch in light-switch example

### PR DESCRIPTION
**Pull request by markus-becker-tridonic-com**
#### Problem

In matter there are 2 types of light switch device types:

- On/Off Light Switch, Dimmer Switch, Color Dimmer Switch, Control Bridge
- Generic Switch

Up to now only the first type was shown in the light-switch example.

#### Change overview

This PR adds GenericSwitch for the nrfconnect light-switch sample.

The GenericSwitch has been added on EP 2 and uses button 3.

The code implements a momentary switch with features MS and MSR.

#### Testing

Tested using chip-tool and DK:

        chip-tool interactive start

Subscribe to the initial-press and short-release

        >>> switch subscribe-event initial-press 1 20 <node_id> 2 --is-urgent true
        >>> switch subscribe-event initial-press 1 20 <node_id> 2 --is-urgent true

### CODE REVIEW

File: .github/.wordlist.txt [hunk 1]
```diff
@@ -841,6 +842,7 @@ MoveToSaturation
 MoveWithOnOff
 MPSL
 MRP
+MSR
```
Comment (andy31415) [inline .github/.wordlist.txt:845]: is MSR a word? I see it used as an acronym, however that is the only use. Could we just not use the acronym or use backticks to escape it from being considered a real word?

File: .github/.wordlist.txt [hunk 2]
```diff
@@ -306,6 +306,7 @@ ctrl
 ctypes
 CurrentHue
 CurrentLevel
+CurrentPosition
```
Comment (andy31415) [inline .github/.wordlist.txt:309]: CurrentPosition is not a word.

File: examples/light-switch-app/nrfconnect/README.md [hunk 3]
```diff
@@ -273,6 +298,11 @@ platform image.
     with 1% increments every 300 milliseconds as long as **Button 2** is
     pressed.
 
+**Button 3** can be used for the following purposes:
+
+-   _Pressed once_ &mdash; Changes the value of the attribute CurrentPosition
```
Comment (andy31415) [inline examples/light-switch-app/nrfconnect/README.md:303]: use backticks to escape things that are not words

**Review by andy31415 — APPROVED** [review]

File: .github/.wordlist.txt (hunk 2 above)
Comment (markus-becker-tridonic-com) [reply .github/.wordlist.txt]: Done.

File: examples/light-switch-app/nrfconnect/README.md (hunk 3 above)
Comment (markus-becker-tridonic-com) [reply examples/light-switch-app/nrfconnect/README.md]: Done.

File: examples/light-switch-app/nrfconnect/main/AppTask.cpp [hunk 4]
```diff
@@ -259,6 +266,11 @@ void AppTask::ButtonReleaseHandler(AppEvent * aEvent)
             sAppTask.CancelTimer(Timer::DimmerTrigger);
             sWasDimmerTriggered = false;
             break;
+        case GENERIC_SWITCH_BUTTON:
+            LOG_INF("GenericSwitch: ShortRelease");
+            sAppTask.StartTimer(Timer::Function, 0);
```
Comment (Damian-Nordic) [inline examples/light-switch-app/nrfconnect/main/AppTask.cpp:271]: Using a zero timeout function seems a bit overcomplicated. Consider adding another event type in `AppEvent.h`.

**Review by Damian-Nordic — CHANGES_REQUESTED** [review]
Thanks. I have one minor comment, but we can also do it as a followup.
Marking as "Request changes" so that our Tech Writer has a chance to review the documentation changes :)

File: examples/light-switch-app/nrfconnect/main/AppTask.cpp (hunk 4 above)
Comment (markus-becker-tridonic-com) [reply examples/light-switch-app/nrfconnect/main/AppTask.cpp:271]: Agree, that I do not need the `StartTimer()`, but I don't think we need another event type (kEventType_Button seems fine). Shall I directly call `LightSwitch::GetInstance().GenericSwitchInitialPress();`/`LightSwitch::GetInstance().GenericSwitchReleasePress();` in `ButtonReleaseHandler()`/`ButtonPushHandler()` or is that the wrong context?

Comment (Damian-Nordic) [reply examples/light-switch-app/nrfconnect/main/AppTask.cpp:271]: You're right, `kEventType_Button` is enough and you can call `GenericSwitchInitialPress/ReleasePress` directly. However, in those methods you can modify the attributes using:
```
DeviceLayer::SystemLayer().ScheduleLambda([] {
  uint8_t newPosition = 1;

  Clusters::Switch::Attributes::CurrentPosition::Set(mLightGenericSwitchEndpointId, newPosition);
  Clusters::SwitchServer::Instance().OnInitialPress(mLightGenericSwitchEndpointId, newPosition);
});
```
to make sure that CHIP data is modified on the CHIP thread.

File: examples/light-switch-app/nrfconnect/main/LightSwitch.cpp [hunk 5]
```diff
@@ -80,3 +86,19 @@ void LightSwitch::DimmerChangeBrightness()
...
+{
+    uint8_t newPosition = 1;
+
+    Clusters::Switch::Attributes::CurrentPosition::Set(mLightGenericSwitchEndpointId, newPosition);
+    Clusters::SwitchServer::Instance().OnInitialPress(mLightGenericSwitchEndpointId, newPosition);
+}
+
+void LightSwitch::GenericSwitchReleasePress()
+{
+    uint8_t previousPosition = 0;
```
Comment (bzbarsky-apple) [inline examples/light-switch-app/nrfconnect/main/LightSwitch.cpp:100]: This is the new position, no?

File: examples/light-switch-app/nrfconnect/main/LightSwitch.cpp [hunk 6]
```diff
@@ -80,3 +86,19 @@ void LightSwitch::DimmerChangeBrightness()
...
+
+    Clusters::Switch::Attributes::CurrentPosition::Set(mLightGenericSwitchEndpointId, previousPosition);
+    Clusters::SwitchServer::Instance().OnShortRelease(mLightGenericSwitchEndpointId, previousPosition);
```
Comment (bzbarsky-apple) [inline examples/light-switch-app/nrfconnect/main/LightSwitch.cpp:103]: This is supposed to be handed the previous position, right?  As in, the value of `CurrentPosition` before we set it here.

File: examples/light-switch-app/light-switch-common/light-switch-app.matter [hunk 7]
```diff
@@ -2000,5 +2034,50 @@ endpoint 1 {
...
+    ram      attribute identifyTime;
+    ram      attribute identifyType;
+    callback attribute generatedCommandList;
+    callback attribute acceptedCommandList;
+    callback attribute attributeList;
+    ram      attribute featureMap;
+    ram      attribute clusterRevision default = 2;
+  }
+
+  server cluster Groups {
```
[the remark this answers is not in the document]
Comment (markus-becker-tridonic-com) [reply examples/light-switch-app/light-switch-common/light-switch-app.matter:2051]: zap GUI automatically added it for Generic Switch device type. Shall I remove it?

File: examples/light-switch-app/light-switch-common/light-switch-app.matter [hunk 8]
```diff
@@ -2000,5 +2034,50 @@ endpoint 1 {
     ram      attribute clusterRevision default = 1;
   }
 }
+endpoint 2 {
+  device type genericswitch = 15;
+  binding cluster Binding;
```
[the remark this answers is not in the document]
Comment (markus-becker-tridonic-com) [reply examples/light-switch-app/light-switch-common/light-switch-app.matter:2039]: zap GUI automatically added it for Generic Switch device type. Shall I remove it?

File: examples/light-switch-app/nrfconnect/main/LightSwitch.cpp (hunk 6 above)
Comment (markus-becker-tridonic-com) [reply examples/light-switch-app/nrfconnect/main/LightSwitch.cpp:103]: See above. This is now the releasedPosition which is 0.

Spec:

> CurrentPosition value 0 SHALL be assigned to the default position of the
> switch: for example the "open" state of a rocker switch, or the "idle" state of a push button switch.

The parameter however is called previousPosition in `void SwitchServer::OnShortRelease(EndpointId endpoint, uint8_t previousPosition)`, which should probably be changed then as well.

File: examples/light-switch-app/light-switch-common/light-switch-app.matter (hunk 8 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply examples/light-switch-app/light-switch-common/light-switch-app.matter:2039]: See above: yes, please.  And file an issue to fix the device type definition, please?

File: examples/light-switch-app/nrfconnect/main/LightSwitch.cpp (hunk 6 above)
Comment (bzbarsky-apple) [reply examples/light-switch-app/nrfconnect/main/LightSwitch.cpp]: No the point is that for a release event the field provided with the event in the spec is the position _before_ the release.  So the parameter name is in fact correct.

File: examples/light-switch-app/nrfconnect/main/LightSwitch.cpp [hunk 9]
```diff
@@ -89,16 +89,20 @@ void LightSwitch::DimmerChangeBrightness()
...
 void LightSwitch::GenericSwitchReleasePress()
 {
-    uint8_t previousPosition = 0;
+    DeviceLayer::SystemLayer().ScheduleLambda([this] {
+        uint8_t releasedPosition = 0;
 
-    Clusters::Switch::Attributes::CurrentPosition::Set(mLightGenericSwitchEndpointId, previousPosition);
-    Clusters::SwitchServer::Instance().OnShortRelease(mLightGenericSwitchEndpointId, previousPosition);
+        Clusters::Switch::Attributes::CurrentPosition::Set(mLightGenericSwitchEndpointId, releasedPosition);
+        Clusters::SwitchServer::Instance().OnShortRelease(mLightGenericSwitchEndpointId, releasedPosition);
```
Comment (bzbarsky-apple) [inline examples/light-switch-app/nrfconnect/main/LightSwitch.cpp:106]: This is not matching the spec as far as I can tell.

File: examples/light-switch-app/nrfconnect/README.md [hunk 10]
```diff
@@ -62,13 +62,38 @@ and [Zephyr RTOS](https://zephyrproject.org/). Visit Matter's
...
+The first type uses Client application clusters (e.g. Level Control) and
+Bindings to send commands to the server clusters. This type of switch is on
+endpoint 1 of this example.
+
+With the second type controllers can use event subscriptions to be informed
+about changes on the Switch server cluster. This type of switch is on endpoint 2
+of this example.
+
+## On/Off Light Switch, Dimmer Switch, Color Dimmer Switch, Control Bridge
+
```
Comment (greg-fer) [inline examples/light-switch-app/nrfconnect/README.md:79]:
```suggestion
In Matter, the following types of light switch devices are available:

-   Group 1: On/Off Light Switch, Dimmer Switch, Color Dimmer Switch, Control Bridge
-   Group 2: Generic Switch

The first type (here referred to as Group 1) uses client application clusters (for example, Level Control) and
bindings to send commands to the server clusters. This type of switch is on the endpoint 1 of this example.

With the second type (Group 2), controllers can use event subscriptions to be informed
about changes on the switch server cluster. This type of switch is on the endpoint 2
of this example.

### Group 1 light switch devices overview

```

File: examples/light-switch-app/nrfconnect/README.md [hunk 11]
```diff
@@ -62,13 +62,38 @@ and [Zephyr RTOS](https://zephyrproject.org/). Visit Matter's
...
 A light switch device is a simple embedded controller, which has the ability to
 control lighting devices, such as light bulbs or LEDs. After commissioning into
 a Matter network, the light switch device does not know what it can control. In
 other words, it has no information about another device being connected to the
 same network. You must provide this information to the light switch through the
 process called binding, which links clusters and endpoints on both devices, so
-that the devices can interact with each other.
+that the devices can interact with each other. This functionality is on endpoint
+1 and triggered by button 2.
```
Comment (greg-fer) [inline examples/light-switch-app/nrfconnect/README.md:87]:
```suggestion
that the devices can interact with each other. This functionality is on the endpoint
1 and triggered by **Button 2**.
```

File: examples/light-switch-app/nrfconnect/README.md [hunk 12]
```diff
@@ -62,13 +62,38 @@ and [Zephyr RTOS](https://zephyrproject.org/). Visit Matter's
...
+
+## Generic Switch
+
+The Generic Switch on endpoint 2 offers the Switch server cluster. It implements
+the features Momentary Switch (`MS`) and Momentary Switch Release (`MSR`), thus
+it will send event notifications `InitialPress` and `ShortRelease` if button 3
+of the DK is pressed and released.
+
+## Common settings
```
Comment (greg-fer) [inline examples/light-switch-app/nrfconnect/README.md:96]:
```suggestion
### Generic Switch

The Generic Switch on the endpoint 2 offers the switch server cluster. It implements
the Momentary Switch (``MS``) and Momentary Switch Release (``MSR``) features.
For this reason, it sends event notifications ``InitialPress`` and ``ShortRelease`` if **Button 3**
of the DK is pressed and released.

### Common example settings
```

File: examples/light-switch-app/nrfconnect/README.md [hunk 13]
```diff
@@ -273,6 +298,11 @@ platform image.
     with 1% increments every 300 milliseconds as long as **Button 2** is
     pressed.
 
+**Button 3** can be used for the following purposes:
+
+-   _Pressed once_ &mdash; Changes the value of the attribute `CurrentPosition`
+    and (if subscribed) will send the event notifications to the controller.
```
Comment (greg-fer) [inline examples/light-switch-app/nrfconnect/README.md:304]:
```suggestion
**Button 3** can be used for the following purposes:

-   _Pressed once_ &mdash; Changes the value of the attribute ``CurrentPosition``
    and (if subscribed) sends the event notifications to the controller.
```

File: examples/light-switch-app/nrfconnect/README.md [hunk 14]
```diff
@@ -708,6 +738,20 @@ use [light switch buttons](#buttons) or
...
+        chip-tool interactive start
+
+Subscribe to the initial-press and short-release
+
+        >>> switch subscribe-event initial-press 1 20 <node_id> 2 --is-urgent true
+        >>> switch subscribe-event initial-press 1 20 <node_id> 2 --is-urgent true
+
+When pressing and releasing button 3, both events should be sent to the
+controller.
+
```
Comment (greg-fer) [inline examples/light-switch-app/nrfconnect/README.md:754]:
```suggestion

To test the Generic Switch, complete the following steps:

1. Enter the interactive mode of the CHIP Tool:

        chip-tool interactive start

2. Subscribe to the ``InitialPress`` and ``ShortRelease`` event notifications:

        >>> switch subscribe-event initial-press 1 20 <node_id> 2 --is-urgent true
        >>> switch subscribe-event initial-press 1 20 <node_id> 2 --is-urgent true

3. Press **Button 3** to test the new configuration.
    When pressing and releasing **Button 3**, both events should be sent to the controller.

```

**Review by greg-fer — CHANGES_REQUESTED** [review]
Couple of comments to ensure style consistency.

File: examples/light-switch-app/light-switch-common/light-switch-app.matter (hunk 8 above)
[the remark this answers is not in the document]
Comment (markus-becker-tridonic-com) [reply examples/light-switch-app/light-switch-common/light-switch-app.matter:2039]: PR: https://github.com/project-chip/connectedhomeip/pull/19881

**Review by Damian-Nordic — APPROVED** [review]
Approving with the doc changes applied.

File: examples/light-switch-app/light-switch-common/light-switch-app.matter (hunk 7 above)
[the remark this answers is not in the document]
Comment (markus-becker-tridonic-com) [reply examples/light-switch-app/light-switch-common/light-switch-app.matter]: removed the unnecessary clusters.

File: examples/light-switch-app/nrfconnect/main/LightSwitch.cpp (hunk 5 above)
Comment (markus-becker-tridonic-com) [reply examples/light-switch-app/nrfconnect/main/LightSwitch.cpp]: Fixed now.

File: examples/light-switch-app/nrfconnect/main/LightSwitch.cpp (hunk 9 above)
Comment (markus-becker-tridonic-com) [reply examples/light-switch-app/nrfconnect/main/LightSwitch.cpp]: Fixed.

**Review by greg-fer — APPROVED** [review]

**Comment by greg-fer** [comment]
Thanks for applying the edits, @markus-becker-tridonic-com !

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/light-switch-app/nrfconnect/main/LightSwitch.cpp (hunk 6 above)
Comment (markus-becker-tridonic-com) [reply examples/light-switch-app/nrfconnect/main/LightSwitch.cpp]: Closing this conversation given @bzbarsky-apple approval. Created issue https://github.com/project-chip/connectedhomeip/issues/19923 for the linux part.

**Comment by markus-becker-tridonic-com** [comment]
Proposals how to fix the misspell error?

It's for keepSubscriptions in 
```
2.  Subscribe to the `InitialPress` and `ShortRelease` event notifications:

        >>> switch subscribe-event initial-press 1 20 <node_id> 2 --is-urgent true --keepSubscriptions true`
```

which is in a code section.

Add to wordlist or has the code section be marked somehow?

**Comment by andy31415** [comment]
> Proposals how to fix the misspell error?
> 
> It's for keepSubscriptions in
> 
> ```
> 2.  Subscribe to the `InitialPress` and `ShortRelease` event notifications:
> 
>         >>> switch subscribe-event initial-press 1 20 <node_id> 2 --is-urgent true --keepSubscriptions true`
> ```
> 
> which is in a code section.
> 
> Add to wordlist or has the code section be marked somehow?

It probably is missing backticks, so it thingks it is words. In the quote I seem to see a missing starting backtick

**Review by andy31415 — APPROVED** [review]

**Pull request merged**